### DISCUSSION TITLE
Collapse and expand the tree with left/right arrows and/or h/l keys?

### DISCUSSION THREAD

**Discussion by cadamsdotcom** [nvim-tree/nvim-tree.lua#2687] [Q&A]
Hi!,
In macOS Finder windows, you can use the left and right arrows to collapse and expand folders and move into and out of the hierarchy. Here's an example: https://www.youtube.com/watch?v=Dx9_RQLqShI&t=401s

It makes for a quick way to descend up and down the folder hierarchy, expanding and collapsing as you go - and it's standard in macOS tree views so I didn't know how much I'd miss it!

Wondering if this behavior can be reproduced in nvim-tree?

A quick glance at the help, and a little play with it out of the box, suggests building this could be possible by inspecting what's under the cursor and deciding how to behave.

On pressing `l` either
* open if the cursor is on a closed folder
* move down if the cursor is on an open folder
* no action if the cursor is on a file

On pressing `h` almost the reverse
* close folder if the cursor is on an open folder
* move to parent folder otherwise

I'm 100% new to neovim and lua so would love some guidance on how to do this!

**Comment by alex-courtis** [comment]
This is all possible using API.

Take a look at the [first recipe](https://github.com/nvim-tree/nvim-tree.lua/wiki/Recipes#h-j-k-l-style-navigation-and-editing) which should cover a lot of your use cases.

**Comment by the-citto** [comment]
been a long time since I set it up (left commented old code,) got here after latest update where `expand_or_collapse` is not available anymore

thank you for the hint @alex-courtis, sorted with 
```lua
on_attach = function (bufnr)
    local opts = { buffer = bufnr }
    api.config.mappings.default_on_attach(bufnr)
    local lefty = function ()
        local node_at_cursor = lib.get_node_at_cursor()
        if (node_at_cursor.name == ".." or node_at_cursor.nodes) and node_at_cursor.open then
            api.node.open.edit()
            -- lib.expand_or_collapse(node_at_cursor)
        else
            api.node.navigate.parent()
        end
    end
    local righty = function ()
        local node_at_cursor = lib.get_node_at_cursor()
        if (node_at_cursor.name == ".." or node_at_cursor.nodes) and not node_at_cursor.open then
            api.node.open.edit()
            -- lib.expand_or_collapse(node_at_cursor)
        end
    end
    vim.keymap.set("n", "h", lefty , opts )
    vim.keymap.set("n", "<Left>", lefty , opts )
    vim.keymap.set("n", "<Right>", righty , opts )
    vim.keymap.set("n", "l", righty , opts )
end,

```
in the recipe it's clearly written that 
```lua
-- expand or collapse folder
api.node.open.edit()
```
but it still took me some time to wrap my head to the fact that `node.open.edit()` actually expands or collapses 😶

**Reply by gegoune** [reply]
If you think documentation could be improved we would really appreciate a PR!

**Reply by the-citto** [reply]
I'll have a look at how to PR a wiki entry :)

in the meanwhile, should this come up on some google search as it did for me, some better comment and removed some redundant code

```lua
M.config = function ()
    local api = require("nvim-tree.api")
    local lib = require("nvim-tree.lib")
    require"nvim-tree".setup {
        on_attach = function (bufnr)
            local opts = { buffer = bufnr }
            api.config.mappings.default_on_attach(bufnr)
            local lefty = function ()
                local node_at_cursor = lib.get_node_at_cursor()
                -- node_at_cursor.nodes to checks if has nodes (is a folder) and .open to check it's open
                if node_at_cursor.nodes and node_at_cursor.open then
                    -- .open.edit() toggles the node (folder) from open to closed
                    api.node.open.edit()
                else
                    -- if instead node_at_cursor is a file or a closed folder, jump to the parent node (folder)
                    api.node.navigate.parent()
                end
            end
            local righty = function ()
                local node_at_cursor = lib.get_node_at_cursor()
                -- node_at_cursor.nodes to checks if has nodes (is a folder) and not .open to check it's closed
                if node_at_cursor.nodes and not node_at_cursor.open then
                    -- .open.edit() toggles the node (folder) from closed to open
                    api.node.open.edit()
                    -- no else here, if it's an open folder or a file do nothing
                end
            end
            vim.keymap.set("n", "h", lefty , opts )
            vim.keymap.set("n", "<Left>", lefty , opts )
            vim.keymap.set("n", "<Right>", righty , opts )
            vim.keymap.set("n", "l", righty , opts )

            -- other on_attach stuff

        end,

        -- other setup stuff

    }
end
```
only left and right (both arrows and `h` `l`) to open/close folders or jump to parent for left

unfortunately it's been some time since I set it up first and I don't remember the sources to give credits

**Reply by alex-courtis** [reply]
> I'll have a look at how to PR a wiki entry :)

No need for a PR; please just follow the style of the surrounding page.

**Reply by the-citto** [reply]
sweet
just had to change my code above as  `lib.get_node_at_cursor()` just stopped working, followed the first recipe :)

added my amended recipe at the [end](https://github.com/nvim-tree/nvim-tree.lua/wiki/Recipes#basic-aria-navigation-arrows-and-vim-keys) right now

**Reply by alex-courtis** [reply]
Many thanks @the-citto !
